flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java

```diff
@@ -0,0 +1,353 @@
+    private final int bufferSize;
+
+    /** The number of buffers to be returned for a single request. */
+    private final int numBuffersPerRequest;
+
+    /**
+     * The maximum number of buffers can be allocated from this buffer pool for a single buffer
+     * requester.
+     */
+    private final int maxBuffersPerRequester;
+
+    /** All available buffers in this buffer pool currently. */
+    @GuardedBy("buffers")
+    private final Queue<MemorySegment> buffers = new ArrayDeque<>();
+
+    /** Account for all the buffers requested per requester. */
+    @GuardedBy("buffers")
+    private final Map<Object, Counter> numBuffersAllocated = new HashMap<>();
+
+    /** Whether this buffer pool has been destroyed or not. */
+    @GuardedBy("buffers")
+    private boolean destroyed;
+
+    /** Whether this buffer pool has been initialized or not. */
+    @GuardedBy("buffers")
+    private boolean initialized;
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize) {
+        // 5 min default buffer request timeout
+        this(totalBytes, bufferSize, 5 * 60 * 1000);
+    }
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize, long requestTimeout) {
+        checkArgument(totalBytes > 0, "Total memory size must be positive.");
+        checkArgument(bufferSize > 0, "Size of buffer must be positive.");
+        checkArgument(requestTimeout > 0, "Request timeout must be positive.");
+
+        this.totalBytes = totalBytes;
+        this.bufferSize = bufferSize;
+        this.requestTimeout = requestTimeout;
+
+        this.numTotalBuffers = (int) Math.min(totalBytes / bufferSize, Integer.MAX_VALUE);
+        this.numBuffersPerRequest =
+                Math.min(numTotalBuffers, Math.max(1, NUM_BYTES_PER_REQUEST / bufferSize));
+        this.maxBuffersPerRequester =
+                Math.max(
+                        4 * numBuffersPerRequest,
+                        numTotalBuffers / BatchShuffleReadIOExecutor.MAX_NUM_THREADS);
+    }
+
+    public long getTotalBytes() {
+        return totalBytes;
+    }
+
+    public long getRequestTimeout() {
+        return requestTimeout;
+    }
+
+    public int getNumBuffersPerRequest() {
+        return numBuffersPerRequest;
+    }
+
+    public int getMaxBuffersPerRequester() {
+        return maxBuffersPerRequester;
+    }
+
+    public int getBufferSize() {
+        return bufferSize;
+    }
+
+    public int getNumTotalBuffers() {
+        return numTotalBuffers;
+    }
+
+    public int getAvailableBuffers() {
+        synchronized (buffers) {
+            return buffers.size();
+        }
+    }
+
+    /** Initializes this buffer pool which allocates all the buffers. */
+    public void initialize() {
+        LOG.info(
+                "Initializing batch shuffle IO buffer pool: numBuffers={}, bufferSize={}.",
+                numTotalBuffers,
+                bufferSize);
+
+        checkState(
+                totalBytes >= bufferSize,
```

Comment:
I add this check to initialize instead of constructor to keep the behavior not changed. Because if one have set the memory segment size to larger than 32M, then after he/she upgrades to the new version, exception will be thrown in the constructor even sort-merge shuffle is not enabled.